File: Stage29/Python/Problem9370.py
import sys
import heapq

input = sys.stdin.readline
INF = 10**15

def dijkstra(start, graph, n):
    dist = [INF] * (n + 1)
    pq = []

    dist[start] = 0
    heapq.heappush(pq, (0, start))

    while pq:
        cost, now = heapq.heappop(pq)

        if cost > dist[now]:
            continue

        for nxt, w in graph[now]:
            nd = cost + w

            if dist[nxt] > nd:
                dist[nxt] = nd
                heapq.heappush(pq, (nd, nxt))

    return dist

def main():
    T = int(input())

    for _ in range(T):
        n, m, t = map(int, input().split())
        s, g, h = map(int, input().split())

        graph = [[] for _ in range(n + 1)]
        gh_cost = 0

        for _ in range(m):
            a, b, d = map(int, input().split())

            graph[a].append((b, d))
            graph[b].append((a, d))

            if (a == g and b == h) or (a == h and b == g):
                gh_cost = d

        candidates = []
        for _ in range(t):
            candidates.append(int(input()))

        candidates.sort()

        distS = dijkstra(s, graph, n)
        distG = dijkstra(g, graph, n)
        distH = dijkstra(h, graph, n)

        for x in candidates:

            path1 = distS[g] + gh_cost + distH[x]
            path2 = distS[h] + gh_cost + distG[x]

            if distS[x] == path1 or distS[x] == path2:
                print(x, end=" ")

        print()

if __name__ == "__main__":
    main()

File: Stage29/Java/Problem9370.java
package Stage29.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Problem9370 {
    static class Edge implements Comparable<Edge> {
        int to, cost;

        Edge(int to, int cost) {
            this.to = to;
            this.cost = cost;
        }

        public int compareTo(Edge o) {
            return this.cost - o.cost;
        }
    }

    static int n, m, t;
    static int s, g, h;
    static final int INF = 1000000000;
    static ArrayList<Edge>[] graph;

    static int[] dijkstra(int start) {
        PriorityQueue<Edge> pq = new PriorityQueue<>();
        int[] dist = new int[n + 1];
        Arrays.fill(dist, INF);

        dist[start] = 0;
        pq.add(new Edge(start, 0));

        while (!pq.isEmpty()) {
            Edge cur = pq.poll();

            if (cur.cost > dist[cur.to]) continue;

            for (Edge next : graph[cur.to]) {
                int nd = cur.cost + next.cost;

                if (dist[next.to] > nd) {
                    dist[next.to] = nd;
                    pq.add(new Edge(next.to, nd));
                }
            }
        }

        return dist;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {

            st = new StringTokenizer(br.readLine());
            n = Integer.parseInt(st.nextToken());
            m = Integer.parseInt(st.nextToken());
            t = Integer.parseInt(st.nextToken());

            st = new StringTokenizer(br.readLine());
            s = Integer.parseInt(st.nextToken());
            g = Integer.parseInt(st.nextToken());
            h = Integer.parseInt(st.nextToken());

            graph = new ArrayList[n + 1];
            for (int i = 1; i <= n; i++) graph[i] = new ArrayList<>();

            int ghCost = 0;

            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());

                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                int d = Integer.parseInt(st.nextToken());

                graph[a].add(new Edge(b, d));
                graph[b].add(new Edge(a, d));

                if ((a == g && b == h) || (a == h && b == g)) {
                    ghCost = d;
                }
            }

            int[] candidate = new int[t];
            for (int i = 0; i < t; i++) {
                candidate[i] = Integer.parseInt(br.readLine());
            }

            Arrays.sort(candidate);

            int[] distS = dijkstra(s);
            int[] distG = dijkstra(g);
            int[] distH = dijkstra(h);

            for (int x : candidate) {

                int path1 = distS[g] + ghCost + distH[x];
                int path2 = distS[h] + ghCost + distG[x];

                if (distS[x] == path1 || distS[x] == path2) {
                    sb.append(x).append(" ");
                }
            }

            sb.append("\n");
        }

        System.out.print(sb);
        br.close();
    }
}